### ISSUE TITLE
Lifecycle hooks do not work with moduleRef.resolve

### ISSUE THREAD

**Issue by Gustav0ar** [nestjs/nest#3872]
## Current behavior
I need to create references to a service with different instances dynamically. When I set the scope of a service as TRANSIENT, it forces me to use moduleRef.resolve which does not fire ApplicationShutdown event when I kill the application. If I inject the service in the constructor it works fine, but then I'll have just one instance, and I need multiple instances.

I need this event because my app connects to different databases in each instance and I need to gracefully close the connections when the application quits

## Input Code
Repo that reproduces the issue: https://github.com/Gustav0ar/nest-shutdown-issue

main.ts
```ts
async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  app.enableShutdownHooks();
  await app.listen(3000);
}
bootstrap();

```

service.ts
```ts
@Injectable({ scope: Scope.TRANSIENT })
export class AppService implements OnApplicationShutdown {
  onApplicationShutdown(signal?: string) {
    console.log('Shutting down: ' + signal);
  }
}
```

controller.ts
```ts
@Controller()
export class AppController implements OnModuleInit {
  private appService: AppService;
  
  constructor(private moduleRef: ModuleRef) {}

  async onModuleInit() {
    this.appService = await this.moduleRef.resolve(AppService);
  }
}
```

## Expected behavior
To call ApplicationShutdown when app exits

**Comment by kamilmysliwiec** [comment]
Dynamically created instances are detached from the application lifecycle. When you create a service using the `resolve()` method, Nest creates a unique context ID + accesses DI container to instantiate your class. However, the framework is not aware of what's the expected lifecycle of this instance. What if you want to create an instance of the service just to perform a specific asynchronous operation/execute business logic and remove it afterwards (by assigning `null` or `undefined` value)? Since this instance doesn't have any references in your app anymore, JS garbage collector should remove it from the memory. If we automatically attached all dynamic instances to the app lifecycle, it wouldn't be possible.

To recap, if you resolve a class dynamically, you should control its lifecycle on your own. Let's say that your class `A` resolves class `B`. Class `A` is now a host class for class `B`. Hence, if `B` lifecycle is tied to `A` lifecycle, you should register hooks in `A` and manually trigger them for `B`. We're tracking the docs here https://github.com/nestjs/docs.nestjs.com/issues/610

**Comment by Gustav0ar** [comment]
@kamilmysliwiec Hi, thanks for the response, I see what you mean. If nest can't keep track of the instance, what is the correct approach to get shutdown event (nodejs or nestjs) when application for some reason (usually uncaught error or a system signal), I need to close down the connection to database I make in those cases, in witch each instance has its own connection.
